Detectors/TPC/simulation/run/convertDigitsToRawZS.cxx

```diff
@@ -106,13 +96,43 @@ void convertDigitsToZSfinal(std::string_view digitsFile, std::string_view output
       writer.registerLink(feeid, cruID, defaultLink, j & 1, fmt::format("{}cru{}.raw", outDir, cruID));
     }
   }
-  for (Long64_t ievent = 0; ievent < treeSim->GetEntries(); ++ievent) {
-    treeSim->GetEntry(ievent);
 
+  treeSim->SetBranchStatus("*", 0);
+  treeSim->SetBranchStatus("TPCDigit_*", 1);
+  for (int iSecBySec = 0; iSecBySec < Sector::MAXSECTOR; ++iSecBySec) {
+    treeSim->ResetBranchAddresses();
     for (int iSec = 0; iSec < Sector::MAXSECTOR; ++iSec) {
-      inputDigits[iSec] = *vDigitsPerSectorCollection[iSec]; //????
+      if (sectorBySector) {
+        iSec = iSecBySec;
+      }
+      vDigitsPerSectorCollection[iSec] = nullptr;
+      treeSim->SetBranchAddress(TString::Format("TPCDigit_%d", iSec), &vDigitsPerSectorCollection[iSec]);
+      if (sectorBySector) {
+        break;
+      }
+    }
+    for (Long64_t ievent = 0; ievent < treeSim->GetEntries(); ++ievent) {
+      DigitArray inputDigits;
+      if (sectorBySector) {
+        treeSim->GetBranch(TString::Format("TPCDigit_%d", iSecBySec))->GetEntry(ievent);
+      } else {
+        treeSim->GetEntry(ievent);
+      }
+
+      for (int iSec = 0; iSec < Sector::MAXSECTOR; ++iSec) {
+        if (sectorBySector) {
+          iSec = iSecBySec;
+        }
+        inputDigits[iSec] = *vDigitsPerSectorCollection[iSec]; //????
+        if (sectorBySector) {
+          break;
```

Comment:
Still, you are reading in loop digits to ``vDigitsPerSectorCollection[iSec]`` and they stay in memory (only the span goes out of scope). I think here you have to do ``vDigitsPerSectorCollection[iSec]->clear()``.
Actually, clear will not help, since the vector will be resized to 0 but the memory will stay allocated. I think in case of sectorBySector you should allways set the address of the only branch you read to the same vector.
And you don't need ``treeSim->ResetBranchAddresses();``

Reply:
But vDigitsPerSectorCollection is owned and managed by root I assume? So I assume I would need to tell ROOT to delete it.

Reply:
OK, so if I understand https://root.cern.ch/doc/master/classTBranchElement.html#a5dba28b97ad8d46ee9e5170018c446cb correctly, ROOT will allocate the vector and I am the owner, so I have to delete it?

Reply:
Yes, root internally will clear the vector when it reads into it, but otherwise, you are the owner